feat: Add GPU and NIC health monitoring to workers for vllm

Ava-A4098 opened the pull request:
#### Overview:
This PR adds GPU and NIC health monitoring capabilities to both prefill and decode workers in the Dynamo distributed LLM serving system. The health monitoring runs concurrently with the main worker endpoints to detect hardware issues and enable proactive failure handling.

#### Details:
- **Added health monitoring integration**: Imported `GPUHealthCheck` and `NicHealthCheck` from `nvidia_resiliency_ext.shared_utils.health_check`
- **Created shared `health_monitor()` function**: Handles both GPU recovery status checks and NIC link health monitoring with configurable intervals
- **Integrated with both worker types**: Added health monitoring to both `init()` (decode worker) and `init_prefill()` (prefill worker) functions
- **Concurrent execution**: Health checks run alongside endpoint serving using `asyncio.gather()`
- **Configurable monitoring**: GPU checks every 30s, NIC checks every 60s by default
- **Local rank support**: NIC monitoring uses worker's data parallel rank for device mapping
- **Failure handling framework**: Added placeholder callbacks for GPU/NIC failure scenarios
- **Future extensibility**: Added TODO for smart interval calculation logic

#### Where should the reviewer start?
- **`main.py`** - Focus on the new `health_monitor()` function (lines ~43-68) and its integration in `init()` and `init_prefill()` functions
- **Import statement** - Verify the `nvidia_resiliency_ext` import is correct
- **`asyncio.gather()` calls** - Check that health monitoring is properly added to both worker types without breaking existing functionality


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Introduced continuous health monitoring for GPU and network interfaces, running in the background alongside service operations.
  * Health checks now log any detected hardware issues, improving system reliability and visibility.

* **Chores**
  * Background health monitoring is now automatically started with both prefill and decode worker services.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Ava-A4098 committed:
feat: add GPU and NIC health monitoring to workers

Signed-off-by: Ava-A4098 <aarnaz@nvidia.com>

Commit:
feat:Add GPU and NIC health monitoring to workers for vllm

Signed-off-by: Ava Arnaz <aarnaz@aarnaz-mlt.client.nvidia.com>

nv-anants committed:
fix: use wheel files for installation in trtllm build (#2372)

nv-tusharma committed:
fix: Copy cuda libraries from devel to runtime stage (#2301)

mohammedabdulwahhab committed:
docs: exporting metrics in k8s (dep-302) (#2271)

whoisj committed:
docs: Update nixl_connect README (#2320)

faradawn committed:
docs: add sglang hicache example and guide (#2388)

nealvaidya committed:
fix: use default batch_size for decode (#2376)

ryanolson committed:
feat: cuda traits and interoperability with external contexts (#2340)

grahamking committed:
fix(preprocessor): Populate model ID in PreprocessedRequest (#2397)

nv-tusharma committed:
fix: Fixes OPS-618 Add Access Token for accessing containers on NGC (#2395)

ishandhanani committed:
docs: add sglang installation variants (#2400)

julienmancuso committed:
feat: add grove multinode support (#2269)

tedzhouhk committed:
feat: standalone profiling script for a given endpoint (#2386)

keivenchang committed:
refactor: rename to system status server for consistency (#2354)

Co-authored-by: Keiven Chang <keivenchang@users.noreply.github.com>

hhzhang16 committed:
feat: profiling PVC updates for better UX (#2402)

hexinw-nvidia commented on components/backends/vllm/src/dynamo/vllm/main.py:
```diff
@@ -212,6 +255,9 @@ async def init(runtime: DistributedRuntime, config: Config):
         handler.kv_publisher = kv_publisher
 
+    # Get local rank for monitoring
+    local_rank = config.engine_args.data_parallel_rank or 0
```
Same question as init_prefill.

krishung5 committed:
fix: Pinned PyTorch version for vLLM container (#2359)

tedzhouhk committed:
feat: support SGLang in pre-deployment sweeping (#2360)

Signed-off-by: Hongkuan Zhou <tedzhouhk@gmail.com>
Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

nv-tusharma committed:
fix: Ensure Login to NGC step works (#2409)

KrishnanPrash committed:
feat: Add frontend support for `min_tokens` and `ignore_eos` (outside of `nvext`) and Structured Output / Guided Decoding (#2380)

Signed-off-by: KrishnanPrash <140860868+KrishnanPrash@users.noreply.github.com>
Co-authored-by: Ryan McCormick <rmccormick@nvidia.com>
Co-authored-by: Ayush Agarwal <ayushag@nvidia.com>

rmccorm4 committed:
fix(build): Pin cuda-python>=12,<13 to avoid trtllm breakage (#2379) (#2396)

tanmayv25 committed:
feat: Add performance sweeps for DeepSeek R1 on GB200 (#2387)

GuanLuo committed:
feat: multi-modal example with vLLM v1 and UX v2 (#2040)

Co-authored-by: krishung5 <krish@nvidia.com>

tedzhouhk committed:
feat: support sglang in sla planner (#2421)

Signed-off-by: Hongkuan Zhou <tedzhouhk@gmail.com>
Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

ishandhanani committed:
chore: update wideep dockerfile for sgl to newest rel (#2422)

grahamking committed:
feat: Allow an endpoint to serve multiple models (#2418)

tedzhouhk committed:
feat: LLM metrics for non-streaming requests in frontend (#2427)

ishandhanani committed:
test: add kv routing test for sglang (#2424)

Signed-off-by: ishandhanani <82981111+ishandhanani@users.noreply.github.com>
Co-authored-by: Neelay Shah <neelays@nvidia.com>

jthomson04 committed:
fix: Add detokenize stream (#2413)

Signed-off-by: jthomson04 <jwillthomson19@gmail.com>

da-x committed:
fix: upgrade cudarc to 0.17.1 (#2341)

Signed-off-by: Dan Aloni <dan.aloni@vastdata.com>
Co-authored-by: Tushar Sharma <tusharma@nvidia.com>

elyasmnvidian committed:
chore: Refactor tool calling for wider support in the future (#2393)

ryan-lempka committed:
feat: enable custom metrics prefix (#2432)

jthomson04 committed:
fix: Fix ETCD and NATS starvation under massive request concurrency (#2384)

Signed-off-by: jthomson04 <jwillthomson19@gmail.com>

rmccorm4 committed:
feat: Add vars to multi-node trtllm slurm scripts to support xP yD deployments (#2429)

tanmayv25 committed:
fix(ci): Reduce the free gpu memory fraction (#2433)

keivenchang committed:
feat: add dynamo environment checker script (#2419)

Co-authored-by: Keiven Chang <keivenchang@users.noreply.github.com>

athreesh committed:
fix: Update gpt-oss.md w/ HF CLI pip install (#2431)

Signed-off-by: Anish <80174047+athreesh@users.noreply.github.com>

JesseStutler committed:
chore: Upgrade volcano dependency version (#2442)

Signed-off-by: jessestutler <jesseincomparable@hotmail.com>

messiaen committed:
feat: logprob handling (#2426)

Signed-off-by: Greg Clark <grclark@nvidia.com>

lyang24 committed:
chore: deprecate sentencepiece tokenizer in lib/llm (#2439)

Signed-off-by: lyang24 <lanqingy93@gmail.com>

krishung5 committed:
build: Pin torch versions and skip lmcache installation for ARM build (#2430)

biswapanda committed:
fix: add instruction to deploy model with inference gateway (#2257)

jorgeantonio21 committed:
feat: add RuntimeConfig to ModelEntry (#2311)

Co-authored-by: Yan Ru Pei <yanrpei@gmail.com>

tzulingk committed:
feat: Add a "model" label to Component metrics (#2389)

jthomson04 committed:
perf: Only compute checksums on debug builds (#2446)

Signed-off-by: jthomson04 <jwillthomson19@gmail.com>

mohammedabdulwahhab committed:
fix: operator defaults (#2398)

Signed-off-by: mohammedabdulwahhab <furkhan324@berkeley.edu>

tanmayv25 committed:
chore: Rename tensorrtllm framework to trtllm (#2435)

julienmancuso committed:
fix: add fields documentation in CRDs (#2447)

keivenchang committed:
feat(metrics): add NATS client metrics to prometheus_metrics_fmt (#2292)

Co-authored-by: Keiven Chang <keivenchang@users.noreply.github.com>

julienmancuso committed:
feat: extract deploymentType as interface (#2405)

keivenchang committed:
fix: remove add_labels in Component (#2458)

grahamking committed:
feat: Support `python -m dynamo.frontend --version` (#2449)

rmccorm4 commented:
@nnshah1 what's the plan for this PR?

julienmancuso committed:
fix: fix examples (#2459)

abrarshivani committed:
feat: Dynamic Endpoint Exposure Based on Model Type (#1447)

nachiketb-nvidia committed:
feat: add basic implementation of reasoning parsers (#2410)

krishung5 committed:
ci: Add vllm multimodal example to pytest (#2451)

richardhuo-nv committed:
build: add zmq lib to support trtllm's UCX connection establishment with zmq (#2381)

Signed-off-by: rihuo <rihuo@nvidia.com>

grahamking committed:
chore(sglang): Warn if --skip-tokenizer-init missing (#2463)

julienmancuso committed:
fix: allow tolerations to be added (#2445)

julienmancuso committed:
feat: inject DGD id in planner env variables (#2460)

saturley-hall committed:
fix: remove kvmanager feature from python 3.12 ai-dynamo-runtime wheel (#2456)

ayushag-nv committed:
chore: Tool call parsers incremental improvements + Model Specific Parsers (#2457)

tedzhouhk committed:
chore: bug fixes in pre-deployment sweeping and vllm_v1 planner; expose num_d/p to k8s metrics (#2454)

tanmayv25 committed:
fix: Fix the Dockerfile name (#2468)

PeaBrane committed:
fix: revert registering runtime config for trtllm backend (deadlock) (#2464)

Signed-off-by: Yan Ru Pei <yanrpei@gmail.com>

ZichengMa committed:
docs: Add LMCache in vLLM README.md (#2462)

Signed-off-by: ZichengMa <zichengma1225@gmail.com>

biswapanda committed:
docs: add instructions to install uv (#2476)

nate-martinez committed:
fix: Fix docker-compose.yml Path in Example README (#2473)

Signed-off-by: Nathan Martinez <96562131+nate-martinez@users.noreply.github.com>

ayushag-nv committed:
chore: enable tool call array parsing (#2466)

rmccorm4 committed:
docs: Bring back some missed release/0.4.0 doc changes, fix broken links, add lychee link checker github action (#2482)

ryan-lempka committed:
fix: use random port assignment for http tests (#2472)

dmitry-tokarev-nv committed:
chore: ATTRIBUTIONS-Go.md (#2355) (#2497)

tedzhouhk committed:
chore: add pre-deployment profiling results on H200 cluster to unblock planner testing (#2495)

Co-authored-by: Hannah Zhang <hannahz@nvidia.com>

keivenchang committed:
fix: replace metrics callback with background scraping to prevent tim… (#2480)

Co-authored-by: Keiven Chang <keivenchang@users.noreply.github.com>

grahamking committed:
feat(http): TLS support (#2492)

julienmancuso committed:
fix: document differences between operator and helm chart DGD install… (#2498)

julienmancuso committed:
Fix: Canonicalize PodGangSet to prevent spurious updates (#2469)

julienmancuso committed:
fix: do not set default for resources (#2471)

ayushag-nv committed:
chore: add support for multi-tool within nested tags (#2501)

nachiketb-nvidia committed:
fix: small build warning fix (#2504)

Signed-off-by: nachiketb-nvidia <nachiketb@nvidia.com>

mohammedabdulwahhab committed:
fix: set frontend defaults (#2484)

atchernych committed:
feat: Add epp-aware gateway integration (#2345)

tedzhouhk committed:
test: add unit test for sla planner's interpolator (#2505)

nnshah1 committed:
chore: updating to use pinned precompiled version (#2489)

ishandhanani committed:
feat(sglang): unify entry point for SGLang backend architecture (#2493)

keivenchang committed:
fix: use tokio spawn / interval.tick(), make nats metric names clearer, fix tests sharing environment variables (temp_env) (#2506)

Co-authored-by: Keiven Chang <keivenchang@users.noreply.github.com>

krishung5 committed:
build: Bump vllm and deepgemm version (#2509)

PeaBrane committed:
feat: router-level request rejection (#2465)

rmccorm4 committed:
fix: Use `/usr` dir prefix when freeing disk space (#2514)

atchernych committed:
feat: skip router when worker id is pre-determined  (#2450)

Co-authored-by: Biswa Panda <biswa.panda@gmail.com>

suzusuzu committed:
feat(frontend): support setting HTTP host via CLI (--http-host) (#2523)

ayushag-nv committed:
feat: add mistral and phi4 tool parser (#2510)

saturley-hall committed:
docs: update support matrix with links to ngc (#2524)

Signed-off-by: Harrison King Saturley-Hall <hsaturleyhal@nvidia.com>

tedzhouhk committed:
feat: add a knob to turn off correction factor in sla planner (#2511)

athreesh committed:
docs: Consolidate documentation and fix redundant headings (#2518)

ryanolson committed:
feat: task scheduler (#2406)

Signed-off-by: Ryan Olson <ryanolson@users.noreply.github.com>

nv-anants committed:
ci: run gitlab triggers on runner group (#2532)

Signed-off-by: Anant Sharma <anants@nvidia.com>

julienmancuso committed:
fix: Implement scaling Grove subresources (#2531)

saturley-hall committed:
fix: add --enable-kvbm doc to build.sh help (#2530)

Signed-off-by: Harrison King Saturley-Hall <hsaturleyhal@nvidia.com>

ryanolson committed:
feat: kvbm + connector (#2258)

Signed-off-by: Ryan Olson <rolson@nvidia.com>
Co-authored-by: Olga Andreeva <oandreeva@nvidia.com>
Co-authored-by: Ziqi Fan <ziqif@nvidia.com>
Co-authored-by: John Thompson <jothomson@nvidia.com>
Co-authored-by: Richard Huo <rihuo@nvidia.com>
Co-authored-by: Zicheng Ma <zichengm@nvidia.com>

dmitry-tokarev-nv committed:
chore: Finish vllm upgrade to 0.10.1 + cleanup (#2528)

tzulingk committed:
fix: Correct the metric name and labels. (#2508)

ZichengMa committed:
docs: Add note for LMCache ARM support (#2535)

Co-authored-by: Dmitry Tokarev <dtokarev@nvidia.com>

tedzhouhk committed:
chore: add hhzhang16 to py codeowner (#2529)

nachiketb-nvidia committed:
chore: Bring async-openai into repo as request starter (#2520)

Co-authored-by: Graham King <grahamk@nvidia.com>

tzulingk committed:
feat:  Rename dynamo_component_concurrent_requests (#2515)

Signed-off-by: Tzu-Ling Kan <tzulingk@nvidia.com>

tanmayv25 committed:
feat: Use health check and improve instructions for perf sweeps (#2423)

jthomson04 committed:
fix: Fix KVBM Guide (#2539)

Signed-off-by: jthomson04 <jwillthomson19@gmail.com>

nv-anants committed:
docs: update grove ref to new url (#2538)

Signed-off-by: Anant Sharma <anants@nvidia.com>

nachiketb-nvidia committed:
chore: remove flatten for chat response types, add reasoning_content (#2543)

Changing the chat completions response objects from structs to types of dynamo_async_openai

Implement aggregator traits for them chat completion structs

add reasoning_content under message and delta message in lib/async-openai

dmitry-tokarev-nv committed:
fix: Dockerfile.sglang - Fixed sglang and dynamo wheels installaiton in ru… (#2537)

dmitry-tokarev-nv committed:
chore: Bumped Dynamo version to 0.4.1 (#2545)

krishung5 committed:
test: Add deepep test for vllm (#2534)

ayushag-nv committed:
feat: added parsers lib (#2542)

PeaBrane committed:
test: lychee offline in pre-merge CI (full check only on push to main) (#2549)

keivenchang committed:
feat: enhance devcontainer configuration and documentation (#2255)

Co-authored-by: Keiven Chang <keivenchang@users.noreply.github.com>

jasonqinzhou committed:
feat: add --tokenizer_path to profile_endpoint.py (#2550)

rmccorm4 committed:
docs: Add health check section to GPT OSS guide (#2556)

PeaBrane committed:
feat: upload/download rust structs directly through NATs object store (#2540)

rmccorm4 committed:
docs: Add pinned install for cuda-python in pip install path (#2553)

julienmancuso committed:
fix: resources in wrong location in vllm disag_router example (#2558)

grahamking committed:
chore: Remove async-openai-macros (#2554)

michaelfeil committed:
feat(request cancellation): pycontext, propagating the `is_stopped` into python land. (#2158)

tedzhouhk committed:
test: add dryrun mode for sla planner (#2557)

Signed-off-by: Hongkuan Zhou <tedzhouhk@gmail.com>
Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

grahamking committed:
fix(hub): Download faster from Hugging Face (#2566)

krishung5 committed:
fix: Increase timeout for deepep test (#2572)

ishandhanani committed:
feat(sglang): allow for multi worker spin up in slurm scripts (#2328)

alec-flowers committed:
feat: allow user to adjust kv_transfer_config (#2517)

ziqifan617 committed:
docs: change sglang hicache example to use hicache-ratio (#2582)

biswapanda committed:
feat: use consistent small models across all deploy examples (#2573)

hhzhang16 committed:
fix: add prometheus to the runtime image (#2565)

tedzhouhk committed:
chore: add planner code owners (#2570)

tedzhouhk committed:
fix: profiling script missing tests when kv cache is tight (#2567)

keivenchang committed:
refactor: share common drt test functions (#2583)

Co-authored-by: Keiven Chang <keivenchang@users.noreply.github.com>

ahinsutime committed:
[bugfix] Removed legacy kustomization reference. Initial commit. (#2043)

Signed-off-by: insukim1994 <insu.kim@moreh.io>
Co-authored-by: insukim1994 <insu.kim@moreh.io>

grahamking committed:
chore: Remove Clone / Sync from DeltaGenerator (#2598)

PeaBrane committed:
feat: register Kv router instance into etcd (#2548)

lyang24 committed:
fix: ensure nats fails fast with jetstream failure (#2590)

Signed-off-by: lyang24 <lanqingy93@gmail.com>

michaelfeil committed:
fix: limit Support for HTTP Body limit in axum server (#2581)

Signed-off-by: Michael Feil <63565275+michaelfeil@users.noreply.github.com>
Co-authored-by: Ryan McCormick <mccormick.codes@gmail.com>

tzulingk committed:
feat: Add model label for vllm backend metrics (#2474)

Co-authored-by: Keiven Chang <keivenchang@users.noreply.github.com>

ishandhanani committed:
fix: `skip-tokenizer-init` by default in sglang (#2595)

ishandhanani committed:
feat: h100 wideep instructions fix (#2594)

aranadive committed:
chore: Update UCX to 1.19.0 release (#2551)

Signed-off-by: Adit Ranadive <aranadive@nvidia.com>

tzulingk committed:
fix: Remove metric_labels from multimodal codes. (#2608)

nachiketb-nvidia committed:
feat: enable basic reasoning parsing of <think> </think> tokens (#2555)

oandreeva-nv committed:
test: Temporary exclude kvbm tests from vllm, nightly, gpu1... (#2611)

tedzhouhk committed:
chore: remove circus dependency (#2602)

nv-anants committed:
docs: add trtllm known issue for al2023 (#2604)

Signed-off-by: Anant Sharma <anants@nvidia.com>

ZichengMa committed:
docs: fix doc for kvbm (#2621)

keivenchang committed:
feat: improve dynamo_check.py messaging & instructions (#2453)

Co-authored-by: Keiven Chang <keivenchang@users.noreply.github.com>

michaelfeil committed:
fix: Httpengine sync-enable-endpoint (#2591)

tzulingk committed:
fix: guard inflight_requests and request_duration from early returns. (#2576)

keivenchang committed:
feat: improve devcontainer setup and documentation (#2578)

Co-authored-by: Keiven Chang <keivenchang@users.noreply.github.com>

alec-flowers committed:
fix: --kv-event-config now respects command line (#2627)

tanmayv25 committed:
fix: Update tensorrt_llm to 1.0.0rc6 (#2606)

biswapanda committed:
fix: prevent crash looping hello world (#2625)

ishandhanani committed:
fix: set functional tests to use qwen (#2617)

biswapanda committed:
feat: add multimodal deployment example for llava based on vllm v1  (#2628)

grahamking committed:
chore(llm): Rename protocols::Endpoint to EndpointId (#2615)

krishung5 committed:
fix: Update pytest markers for deepep test. (#2561)

nv-anants committed:
docs: update trtllm know issue message (#2639)

whoisj committed:
feat: Remove Duplicate Multimodel Nixl Connect Example (#2622)

bhuvan002 committed:
feat: add BaseLogitsProcessor core interface (#2613)

Signed-off-by: Bhuvan Agrawal <11240550+bhuvan002@users.noreply.github.com>

grahamking committed:
fix: Tests now pass with RUST_BACKTRACE set (#2647)

krishung5 committed:
docs: Update supported model in readme for multimodal (#2651)

ziqifan617 committed:
feat: enable dynamo metrics on KVBM (#2626)

ayushag-nv committed:
feat: [vLLM] implement cli args for tool and reasoning parsers (#2619)

richardhuo-nv committed:
fix: [TRTLLM+ LLAMA4 + Eagle 3] Remove the ‘two-models config’ and set the ‘one-model’ solution as the default (#2661)

WaelBKZ committed:
fix: handle missing span_name in logging test assertions (#2665)

tedzhouhk committed:
fix: missing tokenizer args in sla_planner.py (#2667)

grahamking committed:
chore: Rust to 1.89 and edition 2024 (#2659)

biswapanda committed:
fix: hello world yaml and messages (#2634)

keivenchang committed:
fix: move metrics registration to service creation (#2664)

Co-authored-by: Keiven Chang <keivenchang@users.noreply.github.com>

rmccorm4 committed:
fix: Skip checksum tests in release mode since they're not computed (#2669)

julienmancuso committed:
fix: fix manual helm chart (#2648)

julienmancuso committed:
fix: do not fail if backendFramework cannot be detected (#2655)

julienmancuso committed:
fix: fix env vars override (#2640)

julienmancuso committed:
fix: increase shm default size and make it configurable (#2616)

alec-flowers committed:
fix: pytest robustness and parsing error (#2676)

tedzhouhk committed:
fix: correct planner test example after tokenizer fix (#2674)

ziqifan617 committed:
feat: add initial batch of KVBM metrics on match, offload and onboard (#2673)

nachiketb-nvidia committed:
feat: add gpt oss reasoning parser through harmony (#2656)

- couple of refactors
- added a new dependency, openai-harmony
- implemented the gpt oss parser

dmitry-tokarev-nv committed:
chore: vllm 0.10.1.1 (#2641)

Ava-A4098 committed:
feat: add per-GPU & NIC health checks; exit on failure

Signed-off-by: Ava-A4098 <aarnaz@nvidia.com>

Ava-A4098 committed:
feat: add per-GPU & NIC health checks; exit on failure

Signed-off-by: Ava-A4098 <aarnaz@nvidia.com>

Ava-A4098 committed:
feat: add per-GPU & NIC health checks; exit on failure.

Signed-off-by: Ava-A4098 <aarnaz@nvidia.com>

Ava-A4098 commented:
Closing this PR due to messy commit history (186+ commits including unrelated changes).
